Lib/functools.py

```diff
@@ -653,6 +653,48 @@ def cache(user_function, /):
     return lru_cache(maxsize=None)(user_function)
 
 
+################################################################################
+### Function variant registry
+################################################################################
+
+# {key: (max lineno, [variant])}
+_variant_registry = {}
+
+
+def register_variant(key, variant):
+    """Register a function variant."""
+    _variant_registry.setdefault(key, []).append(variant)
+
+
+def get_variants(key):
+    """Get all function variants for the given key."""
+    return _variant_registry.get(key, [])
+
+
+def clear_variants(key=None):
+    """Clear all variants for the given key (or all keys)."""
+    if key is None:
+        _variant_registry.clear()
+    else:
+        _variant_registry.pop(key, None)
+
+
+def get_key_for_callable(func):
+    """Return a key for the given callable.
+
+    This key can be used to register the callable in the variant registry
+    with register_variant() or to get variants for this callable with get_variants().
+
+    If no key can be created, return None.
+    """
+    # classmethod and staticmethod
+    func = getattr(func, "__func__", func)
+    try:
+        return f"{func.__module__}.{func.__qualname__}"
+    except AttributeError:
+        return None
```

Comment:
Under what circumstances do we expect the `except` clause to be reached? This could be better documented, either in the docstring or a comment.

Rather than having a `try/except` clause here to swallow the hypothetical `AttributeError`, and then having to do `if key is not None` checks in both `typing.py` and `functools.py`, perhaps it would be better to advertise in this function's docstring that there's a possibility of this function raising `AttributeError`, take out the `try/except` clause here, and have `try/except` clauses in `typing.py` and `functools.py`?

Reply:
```suggestion
    func = getattr(func, "__func__", func)
    return f"{func.__module__}.{func.__qualname__}"
```

The docstring's been updated and both `functools.py` and `typing.py` account for the possibility of `AttributeError` being raised, but the function still swallows `AttributeError` :)

Reply:
Yes, I forgot to change this part :) I'll push another set of changes tonight.